exporter/elasticexporter/exporter.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package elasticexporter contains an opentelemetry-collector exporter
+// for Elastic APM.
+package elasticexporter
+
+import (
+	"bytes"
+	"compress/zlib"
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	otlptrace "github.com/open-telemetry/opentelemetry-proto/gen/go/trace/v1"
+	"go.elastic.co/apm/transport"
+	"go.elastic.co/fastjson"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/component/componenterror"
+	"go.opentelemetry.io/collector/config/configgrpc"
+	"go.opentelemetry.io/collector/config/configmodels"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/exporter/exporterhelper"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/elasticexporter/internal/translator/elastic"
+)
+
+func newElasticTraceExporter(
+	params component.ExporterCreateParams,
+	cfg configmodels.Exporter,
+) (component.TraceExporter, error) {
+	exporter, err := newElasticExporter(cfg.(*Config), params.Logger)
+	if err != nil {
+		return nil, fmt.Errorf("cannot configure Elastic APM trace exporter: %v", err)
+	}
+	return exporterhelper.NewTraceExporter(cfg, func(ctx context.Context, traces pdata.Traces) (int, error) {
+		var dropped int
+		var errs []error
+		for _, resourceSpans := range pdata.TracesToOtlp(traces) {
```

Comment:
I'm thinking of adding support to receive OTLP directly in the Elastic APM Server later, to ease maintenance and minimise signal loss. That would obviate the need to translate to our wire protocol in opentelemetry-collector.

If/when that happens, I would move the `internal/translator` code out of this repo, and the exporter would be a thin wrapper around the OTLP exporter; it would mostly just be about configuring things like auth headers.

> If not, we'd need to cover the gaps later.

Could you please elaborate on that? What gaps are you referring to? The OTLP exporter will need to do this anyway won't it?

Reply:
> Could you please elaborate on that? What gaps are you referring to? The OTLP exporter will need to do this anyway won't it?

We usually use internal data model `consumer/pdata` not OTLP format directly. So I was thinking that you used the OTLP instead because the internal data interface doesn't work for your use case. In that case we would need to fill the gaps in the interface.

I see your point now. If you wan't to reuse translation from OTLP in the Elastic APM backend later on it make sense to use OTLP instead of internal data model here.

Reply:
Gotcha, thanks for the elaboration. If we don't end up going the route I described, I'll revisit this code and switch over to using the internal data model.

Reply:
We are about to move OTLP generated code to our repository: https://github.com/open-telemetry/opentelemetry-collector/pull/1037. Once we do it this code will no longer compile.

If it happens and you are not here to fix your component quickly we will have no choice but to disable it until you can fix it (according to [the rules](https://github.com/open-telemetry/opentelemetry-collector-contrib#adding-new-components) maintenance of contrib components is responsibility of contributors).

Once we have the generation on our side we may start modifying the code generation to produce more optimal in-memory structures. When we do so we will aim to hide the changes behind the public `pdata` wrappers so that your code does not break. If you use generated OTLP structs directly again this can break your code.

I advise to stay with `pdata` wrappers to avoid this sorts of problems.

Reply:
@tigrannajaryan fair enough, I wasn't aware that was happening. I can see now that it makes more sense to use pdata. I've updated the code accordingly. I suppose we could use that in the Elastic APM backend too.

I've removed the metrics translation for now, and will reinstate it when the switchover to pdata.Metrics is complete.